File: api/v1/schemas/backtest.py
# -*- coding: utf-8 -*-
"""Backtest API schemas."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class BacktestRunRequest(BaseModel):
    code: Optional[str] = Field(None, description="仅回测指定股票")
    force: bool = Field(False, description="强制重新计算")
    eval_window_days: Optional[int] = Field(None, ge=1, le=120, description="评估窗口（交易日数）")
    min_age_days: Optional[int] = Field(None, ge=0, le=365, description="分析记录最小天龄（0=不限）")
    limit: int = Field(200, ge=1, le=2000, description="最多处理的分析记录数")
    strategies: Optional[List[str]] = Field(
        None,
        deprecated=True,
        description="Deprecated since v3: backtest now replays each AI plan as-is, no strategy override.",
    )


class BacktestRunResponse(BaseModel):
    processed: int = Field(..., description="候选记录数")
    saved: int = Field(..., description="写入回测结果数")
    completed: int = Field(..., description="完成回测数")
    insufficient: int = Field(..., description="数据不足数")
    errors: int = Field(..., description="错误数")


class BacktestResultItem(BaseModel):
    analysis_history_id: int
    code: str
    name: str = ""
    analysis_date: Optional[str] = None
    eval_window_days: int
    eval_status: str
    evaluated_at: Optional[str] = None
    operation_advice: Optional[str] = None
    position_recommendation: Optional[str] = None
    start_price: Optional[float] = None
    end_close: Optional[float] = None
    max_high: Optional[float] = None
    min_low: Optional[float] = None
    stock_return_pct: Optional[float] = None
    direction_expected: Optional[str] = None
    direction_correct: Optional[bool] = None
    outcome: Optional[str] = None
    stop_loss: Optional[float] = None
    take_profit: Optional[float] = None
    hit_stop_loss: Optional[bool] = None
    hit_take_profit: Optional[bool] = None
    first_hit: Optional[str] = None
    first_hit_date: Optional[str] = None
    first_hit_trading_days: Optional[int] = None
    simulated_entry_price: Optional[float] = None
    simulated_exit_price: Optional[float] = None
    simulated_exit_reason: Optional[str] = None
    simulated_return_pct: Optional[float] = None
    # v2: System-computed signal snapshot + sim diagnostics
    signal_score_at_eval: Optional[int] = None
    buy_signal_at_eval: Optional[str] = None  # STRONG_BUY/BUY/HOLD/AVOID/STRONG_AVOID
    market_environment_at_eval: Optional[str] = None
    strategy_id: Optional[str] = None
    risk_reward_at_eval: Optional[float] = None
    position_pct_at_eval: Optional[float] = None
    trend_score_at_eval: Optional[int] = None
    bias_score_at_eval: Optional[int] = None
    volume_score_at_eval: Optional[int] = None
    support_score_at_eval: Optional[int] = None
    macd_score_at_eval: Optional[int] = None
    rsi_score_at_eval: Optional[int] = None
    capital_flow_score_at_eval: Optional[int] = None
    exit_reason: Optional[str] = None  # stop_loss / take_profit / time_exit / not_filled / not_filled_limit_up / stop_loss_ambiguous / cash
    hold_days: Optional[int] = None
    # v3: AI-plan execution
    entry_status: Optional[str] = None  # filled / not_filled / not_filled_limit_up
    r_multiple: Optional[float] = None
    mae_pct: Optional[float] = None
    mfe_pct: Optional[float] = None


class BacktestResultsResponse(BaseModel):
    total: int
    page: int
    limit: int
    items: List[BacktestResultItem] = Field(default_factory=list)


class PerformanceMetrics(BaseModel):
    scope: str
    code: Optional[str] = None
    eval_window_days: int
    computed_at: Optional[str] = None

    total_evaluations: int
    completed_count: int
    insufficient_count: int
    long_count: int
    cash_count: int
    win_count: int
    loss_count: int
    neutral_count: int

    direction_accuracy_pct: Optional[float] = None
    win_rate_pct: Optional[float] = None
    neutral_rate_pct: Optional[float] = None
    avg_stock_return_pct: Optional[float] = None
    avg_simulated_return_pct: Optional[float] = None

    stop_loss_trigger_rate: Optional[float] = None
    take_profit_trigger_rate: Optional[float] = None
    ambiguous_rate: Optional[float] = None
    avg_days_to_first_hit: Optional[float] = None

    # v3: AI-plan execution metrics
    fill_rate_pct: Optional[float] = None
    filled_count: int = 0
    not_filled_count: int = 0
    not_filled_limit_up_count: int = 0
    trade_win_rate_pct: Optional[float] = None
    expectancy_pct: Optional[float] = None
    avg_r_multiple: Optional[float] = None
    profit_factor: Optional[float] = None
    max_drawdown_pct: Optional[float] = None
    avg_mae_pct: Optional[float] = None
    avg_mfe_pct: Optional[float] = None
    ambiguous_count: int = 0

    diagnostics: Dict[str, Any] = Field(default_factory=dict)
    signal_breakdown: Dict[str, Any] = Field(default_factory=dict)
    score_bucket_breakdown: Dict[str, Any] = Field(default_factory=dict)
    exit_reason_breakdown: Dict[str, Any] = Field(default_factory=dict)
    regime_breakdown: Dict[str, Any] = Field(default_factory=dict)
    strategy_breakdown: Dict[str, Any] = Field(default_factory=dict)

File: apps/dsa-web/src/types/backtest.ts
/**
 * Backtest API type definitions
 * Mirrors api/v1/schemas/backtest.py
 */

// ============ Request / Response ============

export interface BacktestRunRequest {
  code?: string;
  force?: boolean;
  evalWindowDays?: number;
  minAgeDays?: number;
  limit?: number;
}

export interface BacktestRunResponse {
  processed: number;
  saved: number;
  completed: number;
  insufficient: number;
  errors: number;
}

// ============ Result Item ============

export interface BacktestResultItem {
  analysisHistoryId: number;
  code: string;
  name?: string;
  analysisDate?: string;
  evalWindowDays: number;
  evalStatus: string;
  evaluatedAt?: string;
  operationAdvice?: string;
  positionRecommendation?: string;
  startPrice?: number;
  endClose?: number;
  maxHigh?: number;
  minLow?: number;
  stockReturnPct?: number;
  directionExpected?: string;
  directionCorrect?: boolean;
  outcome?: string;
  stopLoss?: number;
  takeProfit?: number;
  hitStopLoss?: boolean;
  hitTakeProfit?: boolean;
  firstHit?: string;
  firstHitDate?: string;
  firstHitTradingDays?: number;
  simulatedEntryPrice?: number;
  simulatedExitPrice?: number;
  simulatedExitReason?: string;
  simulatedReturnPct?: number;
  // v2: System-computed signal snapshot + sim diagnostics
  signalScoreAtEval?: number;
  buySignalAtEval?: string;  // STRONG_BUY/BUY/HOLD/AVOID/STRONG_AVOID
  marketEnvironmentAtEval?: string;
  strategyId?: string;
  riskRewardAtEval?: number;
  positionPctAtEval?: number;
  trendScoreAtEval?: number;
  biasScoreAtEval?: number;
  volumeScoreAtEval?: number;
  supportScoreAtEval?: number;
  macdScoreAtEval?: number;
  rsiScoreAtEval?: number;
  capitalFlowScoreAtEval?: number;
  exitReason?: string;
  holdDays?: number;
  // v3: AI-plan execution
  entryStatus?: string;
  rMultiple?: number;
  maePct?: number;
  mfePct?: number;
}

export interface BacktestResultsResponse {
  total: number;
  page: number;
  limit: number;
  items: BacktestResultItem[];
}

// ============ Performance Metrics ============

export interface PerformanceMetrics {
  scope: string;
  code?: string;
  evalWindowDays: number;
  computedAt?: string;

  totalEvaluations: number;
  completedCount: number;
  insufficientCount: number;
  longCount: number;
  cashCount: number;
  winCount: number;
  lossCount: number;
  neutralCount: number;

  directionAccuracyPct?: number;
  winRatePct?: number;
  neutralRatePct?: number;
  avgStockReturnPct?: number;
  avgSimulatedReturnPct?: number;

  stopLossTriggerRate?: number;
  takeProfitTriggerRate?: number;
  ambiguousRate?: number;
  avgDaysToFirstHit?: number;

  // v3 AI-plan execution metrics
  fillRatePct?: number;
  filledCount: number;
  notFilledCount: number;
  notFilledLimitUpCount: number;
  tradeWinRatePct?: number;
  expectancyPct?: number;
  avgRMultiple?: number;
  profitFactor?: number;
  maxDrawdownPct?: number;
  avgMaePct?: number;
  avgMfePct?: number;
  ambiguousCount: number;

  diagnostics: Record<string, unknown>;
  signalBreakdown?: Record<string, BreakdownBucket>;
  scoreBucketBreakdown?: Record<string, BreakdownBucket>;
  riskRewardBreakdown?: Record<string, BreakdownBucket>;
  exitReasonBreakdown?: Record<string, BreakdownBucket>;
  regimeBreakdown?: Record<string, BreakdownBucket>;
  strategyBreakdown?: Record<string, BreakdownBucket>;
}

export interface BreakdownBucket {
  total: number;
  win: number;
  loss: number;
  neutral: number;
  win_rate_pct?: number | null;
}
